[receiver/postgresql] Flaky test TestPostgreSQLIntegration/without_resource_attributes

Flaky integration test https://github.com/open-telemetry/opentelemetry-collector-contrib/actions/runs/3370453453/jobs/5591988496

```
=== RUN   TestPostgreSQLIntegration
2022/11/01 15:55:56 github.com/testcontainers/testcontainers-go - Connected to docker: 
  Server Version: 20.10.18+azure-2
  API Version: 1.41
  Operating System: Ubuntu 20.04.5 LTS
  Total Memory: 6[944](https://github.com/open-telemetry/opentelemetry-collector-contrib/actions/runs/3370453453/jobs/5591988496#step:5:945) MB
2022/11/01 15:55:56 Starting container id: 24ae72f58407 image: docker.io/testcontainers/ryuk:0.3.4
2022/11/01 15:55:57 Waiting for container id 24ae72f58407 image: docker.io/testcontainers/ryuk:0.3.4
2022/11/01 15:55:57 Container is ready id: 24ae72f58407 image: docker.io/testcontainers/ryuk:0.3.4
2022/11/01 15:56:06 Starting container id: e354be440991 image: c712e5e9-4c1e-423a-9401-0cb423535604:3a7d7256-359a-497a-b4c9-a59e1fb5017a
2022/11/01 15:56:06 Waiting for container id e354be440991 image: c712e5e9-4c1e-423a-9401-0cb423535604:3a7d7256-359a-497a-b4c9-a59e1fb5017a
2022/11/01 15:56:07 Container is ready id: e354be440991 image: c712e5e9-4c1e-423a-9401-0cb423535604:3a7d7256-359a-497a-b4c9-a59e1fb5017a
=== RUN   TestPostgreSQLIntegration/single_db
=== RUN   TestPostgreSQLIntegration/multi_db
=== RUN   TestPostgreSQLIntegration/all_db
=== RUN   TestPostgreSQLIntegration/without_resource_attributes
    integration_test.go:156: 
        	Error Trace:	/home/runner/work/opentelemetry-collector-contrib/opentelemetry-collector-contrib/receiver/postgresqlreceiver/integration_test.go:156
        	Error:      	Received unexpected error:
        	            	the following errors occurred:
        	            	 -  datapoints for metric: `postgresql.backends`, do not match expected
        	            	 -  number of datapoints does not match expected: 1, actual: 2
        	Test:       	TestPostgreSQLIntegration/without_resource_attributes
--- FAIL: TestPostgreSQLIntegration (54.61s)
    --- PASS: TestPostgreSQLIntegration/single_db (11.00s)
    --- PASS: TestPostgreSQLIntegration/multi_db (11.01s)
    --- PASS: TestPostgreSQLIntegration/all_db (11.01s)
    --- FAIL: TestPostgreSQLIntegration/without_resource_attributes (11.01s)
FAIL
coverage: 73.3% of statements
FAIL	github.com/open-telemetry/opentelemetry-collector-contrib/receiver/postgresqlreceiver	54.652s
```

--- a/internal/scrapertest/compare_test.go
+++ b/internal/scrapertest/compare_test.go
@@ -335,6 +335,32 @@ func TestCompareMetrics(t *testing.T) {
 			},
 		},
 		{
+			name: "ignore-subsequent-data-points-all",
+			compareOptions: []CompareOption{
+				IgnoreSubsequentDataPoints(),
+			},
+			withoutOptions: expectation{
+				err: multierr.Combine(
+					errors.New("datapoints for metric: `sum.one`, do not match expected"),
+					errors.New("number of datapoints does not match expected: 1, actual: 2"),
+				),
+				reason: "An unpredictable data point value will cause failures if not ignored.",
+			},
+		},
+		{
+			name: "ignore-subsequent-data-points-one",
+			compareOptions: []CompareOption{
+				IgnoreSubsequentDataPoints("sum.one"),
+			},
+			withoutOptions: expectation{
+				err: multierr.Combine(
+					errors.New("datapoints for metric: `sum.one`, do not match expected"),
+					errors.New("number of datapoints does not match expected: 1, actual: 2"),
+				),
+				reason: "An unpredictable data point value will cause failures if not ignored.",
+			},
+		},
+		{
 			name: "ignore-single-metric",
 			compareOptions: []CompareOption{
 				IgnoreMetricValues("sum.two"),

--- a/internal/scrapertest/mask.go
+++ b/internal/scrapertest/mask.go
@@ -170,3 +170,44 @@ func maskResourceAttributeValue(metrics pmetric.Metrics, opt ignoreResourceAttri
 		}
 	}
 }
+
+// IgnoreSubsequentDataPoints is a CompareOption that ignores data points after the first
+func IgnoreSubsequentDataPoints(metricNames ...string) CompareOption {
+	return ignoreSubsequentDataPoints{
+		metricNames: metricNames,
+	}
+}
+
+type ignoreSubsequentDataPoints struct {
+	metricNames []string
+}
+
+func (opt ignoreSubsequentDataPoints) apply(expected, actual pmetric.Metrics) {
+	maskSubsequentDataPoints(expected, opt.metricNames...)
+	maskSubsequentDataPoints(actual, opt.metricNames...)
+}
+
+func maskSubsequentDataPoints(metrics pmetric.Metrics, metricNames ...string) {
+	metricNameSet := make(map[string]bool, len(metricNames))
+	for _, metricName := range metricNames {
+		metricNameSet[metricName] = true
+	}
+
+	rms := metrics.ResourceMetrics()
+	for i := 0; i < rms.Len(); i++ {
+		sms := rms.At(i).ScopeMetrics()
+		for j := 0; j < sms.Len(); j++ {
+			ms := sms.At(j).Metrics()
+			for k := 0; k < ms.Len(); k++ {
+				if len(metricNames) == 0 || metricNameSet[ms.At(k).Name()] {
+					dps := getDataPointSlice(ms.At(k))
+					n := 0
+					dps.RemoveIf(func(pmetric.NumberDataPoint) bool {
+						n++
+						return n > 1
+					})
+				}
+			}
+		}
+	}
+}

--- a/internal/scrapertest/testdata/ignore-subsequent-data-points-all/actual.json
+++ b/internal/scrapertest/testdata/ignore-subsequent-data-points-all/actual.json
@@ -1,0 +1,38 @@
+{
+   "resourceMetrics": [
+      {
+         "scopeMetrics": [
+            {
+               "metrics": [
+                  {
+                     "name": "sum.one",
+                     "sum": {
+                        "dataPoints": [
+                           {
+                              "asInt": 123
+                           },
+                           {
+                              "asInt": 999
+                           }
+                        ]
+                     }
+                  },
+                  {
+                     "name": "sum.two",
+                     "sum": {
+                        "dataPoints": [
+                           {
+                              "asInt": 234
+                           },
+                           {
+                              "asInt": 999
+                           }
+                        ]
+                     }
+                  }
+               ]
+            }
+         ]
+      }
+   ]
+}

--- a/internal/scrapertest/testdata/ignore-subsequent-data-points-all/expected.json
+++ b/internal/scrapertest/testdata/ignore-subsequent-data-points-all/expected.json
@@ -1,0 +1,32 @@
+{
+   "resourceMetrics": [
+      {
+         "scopeMetrics": [
+            {
+               "metrics": [
+                  {
+                     "name": "sum.one",
+                     "sum": {
+                        "dataPoints": [
+                           {
+                              "asInt": 123
+                           }
+                        ]
+                     }
+                  },
+                  {
+                     "name": "sum.two",
+                     "sum": {
+                        "dataPoints": [
+                           {
+                              "asInt": 234
+                           }
+                        ]
+                     }
+                  }
+               ]
+            }
+         ]
+      }
+   ]
+}

--- a/internal/scrapertest/testdata/ignore-subsequent-data-points-one/actual.json
+++ b/internal/scrapertest/testdata/ignore-subsequent-data-points-one/actual.json
@@ -1,0 +1,25 @@
+{
+   "resourceMetrics": [
+      {
+         "scopeMetrics": [
+            {
+               "metrics": [
+                  {
+                     "name": "sum.one",
+                     "sum": {
+                        "dataPoints": [
+                           {
+                              "asInt": 123
+                           },
+                           {
+                              "asInt": 999
+                           }
+                        ]
+                     }
+                  }
+               ]
+            }
+         ]
+      }
+   ]
+}

--- a/internal/scrapertest/testdata/ignore-subsequent-data-points-one/expected.json
+++ b/internal/scrapertest/testdata/ignore-subsequent-data-points-one/expected.json
@@ -1,0 +1,22 @@
+{
+   "resourceMetrics": [
+      {
+         "scopeMetrics": [
+            {
+               "metrics": [
+                  {
+                     "name": "sum.one",
+                     "sum": {
+                        "dataPoints": [
+                           {
+                              "asInt": 123
+                           }
+                        ]
+                     }
+                  }
+               ]
+            }
+         ]
+      }
+   ]
+}

--- a/receiver/postgresqlreceiver/integration_test.go
+++ b/receiver/postgresqlreceiver/integration_test.go
@@ -153,7 +153,11 @@ func TestPostgreSQLIntegration(t *testing.T) {
 
 			actualMetrics := consumer.AllMetrics()[0]
 
-			require.NoError(t, scrapertest.CompareMetrics(expectedMetrics, actualMetrics, scrapertest.IgnoreMetricValues()))
+			require.NoError(t, scrapertest.CompareMetrics(
+				expectedMetrics, actualMetrics,
+				scrapertest.IgnoreMetricValues(),
+				scrapertest.IgnoreSubsequentDataPoints("postgresql.backends"),
+			))
 		})
 	}
 }

